rancher/terraform-provider-rancher2
# Implement retry logic to enforce timeouts

## Issue: rancher/rancher#37161 

  
## Problem

Currently, some resources in the Terraform Rancher2 provider have timeouts declared and described in documentation, but do not consume these timeout values for their operations. While most of these operations are expected to be short and finish within a reasonable amount of time (often within seconds), it does seem that they might end up hanging indefinitely should something go wrong in the backend.

## Solution

Use the `resource.Retry(time.Duration, RetryFunc)` function to enclose blocks that we would want to have timeouts applied. As a bonus, retry functionality may be easily implemented in the future, if deemed necessary.

As outlined by @eliyamlevy in rancher/rancher#37161, the following resources will have this change applied:
* [x] `rancher2_certificate` (create, update, delete)
* [x] ~`rancher2_cluster_sync` (update, delete)~ _update and delete are no-ops for this resource, and create already has timeouts implemented_
* [x] `rancher2_cluster_template` (create, update, delete)
* [x] `rancher2_global_role` (create, update, delete)
* [x] `rancher2_registry` (create, update, delete)
* [x] `rancher2_role_template` (create, update, delete)
* [x] `rancher2_secret` (create, update, delete)
* [x] `rancher2_token` (create, delete)

## Testing

 
## Engineering Testing
### Manual Testing

* Tested all changed resources manually with a newly-built binary for creation, update, and deletion to make sure no changes in behaviour happened.
 
### Automated Testing

* Acceptance test suite only.
 
## QA Testing Considerations

  
### Regressions Considerations
